File: r/2021/01/main.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
    stop("Must provide input file name")
}

read_input <- function(fname) {
    as.numeric(readLines(fname, warn = FALSE))
}

part_1 <- function(dat) {
    total <- 0
    for (i in seq_len(length(dat) - 1)){
        if (dat[i] < dat[i + 1]) {
            total <- total + 1
        }
    }
    total
}

part_2 <- function(dat) {
    total <- 0
    for (i in seq_len(length(dat) - 3)) {
        if (sum(dat[(i + 1) : (i + 3)]) > sum(dat[i : (i + 2)])) {
            total <- total + 1
        }
    }
    total
}

dat <- read_input(args[1])
print(part_1(dat))
print(part_2(dat))

File: python/2021/01/main.py
import os
import sys
from pathlib import Path
from typing import List


def read_input(fname: str) -> List[int]:
    with open(fname, "r") as f:
        try:
            lines = [int(i) for i in f.readlines()]
        except ValueError:
            raise Exception("Bad input")
    return lines


def part_1(lines: List[int]) -> int:
    return sum(lines[i+1] > lines[i] for i in range(len(lines)-1))


def part_2(lines: List[int]) -> int:
    return sum(sum(lines[i+1:i+4]) > sum(lines[i:i+3])  for i in range(len(lines)-4))


if __name__ == "__main__":
    print(os.path.dirname(__file__))
    p = Path(__file__)
    print(p.absolute())
    data = read_input(sys.argv[1])
    print(part_1(data))
    print(part_2(data))
